src/VisualStudio/TestUtilities/IntegrationHost.cs

```diff
@@ -0,0 +1,522 @@
+                foreach (Window window in _dte.Windows)
+                {
+                    if (window.Caption.Equals(windowTitle, StringComparison.OrdinalIgnoreCase))
+                    {
+                        dteWindow = window;
+                        return true;
+                    }
+                }
+                return false;
+            }, expectedResult: true, timeout: timeout);
+
+            return dteWindow;
+        }
+
+        internal T ExecuteOnHostProcess<T>(Type type, string methodName, BindingFlags bindingFlags, params object[] parameters)
+             => ExecuteOnHostProcess<T>(type.Assembly.Location, type.FullName, methodName, bindingFlags, parameters);
+
+        internal T ExecuteOnHostProcess<T>(string assemblyFilePath, string typeFullName, string methodName, BindingFlags bindingFlags, params object[] parameters)
+        {
+            var objectUri = _service.Execute(assemblyFilePath, typeFullName, methodName, bindingFlags, parameters);
+
+            if (string.IsNullOrWhiteSpace(objectUri))
+            {
+                return default(T);
+            }
+
+            return (T)(Activator.GetObject(typeof(T), $"{_serviceUri}/{objectUri}"));
+        }
+
+        internal void WaitForDteCommandAvailability(string command, int timeout = IntegrationHelper.DefaultTimeout)
+            => IntegrationHelper.WaitForResult(() => Dte.Commands.Item(command).IsAvailable, expectedResult: true, timeout: timeout);
+
+        protected virtual void Dispose(bool disposing)
+        {
+            _requireNewInstance |= (!disposing);
+            Cleanup();
+        }
+
+        private void CleanupDte()
+        {
+            // DTE can still cause a failure or crash during cleanup
+            try
+            {
+                if (_dte == null)
+                {
+                    return;
+                }
+
+                _dte.Documents.CloseAll(vsSaveChanges.vsSaveChangesNo);
+
+                if (_dte.Solution != null)
+                {
+                    // Save the full path to each project in the solution. This is
+                    // so we can cleanup any folders after the solution is closed.
+                    var projectFiles = new List<string>();
+
+                    foreach (Project project in _dte.Solution.Projects)
+                    {
+                        projectFiles.Add(project.FullName);
+                    }
+
+                    // Save the full path to the solution. This is so we can
+                    // cleanup any folders after the solution is closed.
+                    var solutionFile = _dte.Solution.FullName;
+
+                    _dte.Solution.Close(SaveFirst: false);
+
+                    foreach (var projectFile in projectFiles)
+                    {
+                        var projectFolder = Path.GetDirectoryName(projectFile);
+                        _requireNewInstance |= IntegrationHelper.TryDeleteDirectoryRecursively(projectFolder);
```

Comment:
The more I see this flag used the less clear the intention of it.  I think the name either needs to be cleared up or we should add a comment to the declaration that explains the intent of this field.